### ISSUE TITLE
Adding camera functionality

### ISSUE THREAD

**Issue by WardBenjamin** [craftworkgames/MonoGame.Extended#1]
Hey Dylan, 

I'm interested in contributing a few utility methods for the Camera2D class, mainly some methods to check if a point or rectangle is on screen, but possibly a few others. Your README specified that I should discuss what you think would be good contributions, so here's the discussion :wink:

**Comment by craftworkgames** [comment]
Hi @WardBenjamin 

Welcome to the project. I started it less than 2 weeks ago, so it's great to see some interest in it already :smiley: I'm still trying to figure out how everything is going work around here. Since this is the first request, let's just play it by ear and see how it goes.

For smallish changes like this I don't think we need to discuss it forever. Maybe just list some bullet points on what changes you want to make? An example of how it might be used? Or a code snippet if you already have one?

**Comment by craftworkgames** [comment]
The other thing I find helpful is to take a look at how other game frameworks do things. For example, the [LibGDX camera](http://libgdx.badlogicgames.com/nightlies/docs/api/com/badlogic/gdx/graphics/Camera.html) has a lot of good functionality we could consider bringing over.

I don't think it's a good idea to copy the API exactly. Mostly because I want MonoGame.Extended to feel like it's part of MonoGame. In other words, make it more C-sharpy.

Anyway, it looks like the LibGDX camera exposes a Frustrum property for this kind of thing. Maybe that's a good way to go. What do you think?

**Comment by WardBenjamin** [comment]
I definitely think that that is the way to go. I did some research into improving my camera setup very recently and am in the process of integrating a Monogame BoundingFrustrum into it (currently I have a rectangle representing the viewport, which does not support rotation). 

Checking the code, I see that we currently have the following functionality already implemented:
- `Vector2 Position`
- `float Rotation`
- `float Zoom`
- `Vector2 Origin` 
- `void Move(Vector2 direction)`
- `void LookAt(Vector2 position)`
- `Vector2 WorldToScreen(Vector2 worldPosition)`
- `Vector2 ScreenToWorld(Vector2 screenPosition)`
- `Matrix GetViewMatrix(Vector2 parallaxFactor)`
- `Matrix GetViewMatrix()`

Some obvious (for me) additions to this include:
- `BoundingFrustrum Frustrum`
- Methods to check if a Point is in the viewable area (aka the frustrum), with Rectangle and Vector2/3 overloads (possibly more)
- `Matrix InverseViewMatrix()`

That is definitely not a comprehensive list, more should probably be added. We should also at some point discuss a 3D camera, there's a lot of functionality there to implement. 

FYI, I found this via your new blog (not the old craftworkgames one, which I have also seen) when googling around for Camera info. Keep it up! You could probably write an article about contributions :wink:

**Comment by craftworkgames** [comment]
Yep we should definitely add the [BoundingFrustum](https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/BoundingFrustum.cs) to the camera. 

The naming convention I've been using would be to have a `GetBoundingFrustum` method instead of a property, to imply that it does some heavy calculations. That's why I've gone with `GetViewMatrix` instead of a `ViewMatrix` property. Does that make sense?  It's only early in the project so I'm open to discussion if you think it's a bad way to go?

A `GetInverseViewMatrix` also makes sense.

The other methods you're proposing sound like:
 - `Contains(Rectangle rectangle)`
 - `Contains(Point point)` 
 - `Contains(Vector2 point)`
 - etc

At first I thought we wouldn't need those because the `BoundingFrustum` already provides some `Contains` overloads. However, taking a closer look, it doesn't cover many, so I guess putting them directly on the camera class makes sense.

As the project grows we'll also need to deal with testing all this stuff. Perhaps introduce unit tests or at least a *demo* project to show off all the features. That means we'll need a 2D and 3D demo if we want to tackle a 3D camera too.

I only just started the new blog a few weeks ago. The old blog never really took off, but the new one is already working out nicely. At some point I'll have to clean up my websites and figure out what I'm going to with them all. Thanks for letting me know :smile: 

Oh, btw.. You can go ahead and start implementing this when you're ready. No need to discuss every little detail but if you run into some issues let me know.

**Comment by WardBenjamin** [comment]
Alright, I'm going to go ahead and get started. I'll definitely use the same conventions as you had before, they're fine. I personally use a ViewMatrix property but the method works just as well, and makes a lot of sense when you bring up the computational complexity behind it (compared, to say, a float).

I definitely think we should explore some kind of unit testing or something; I have some experience with Travis CI for testing, and a small amount writing unit tests with NUnit for my MonoGame pull request. A demo is also definitely a must, it's nice to have a few examples. Perhaps once the functionality is in place, we add some simple demos in a MonoGame.Extended.Demos repo?

One last thing: the semantics behind the `Contains` methods. I previously used `IsOnScreen` for my implementation, though that's fairly clumsy compared to the more elegant `Contains`. What's your opinion on this?

**Comment by craftworkgames** [comment]
Sounds great. I've already got a build system setup using a team city server. I added a link to the readme tonight. I'm not fussed what we use for the testing framework. NUnit, xUnit or MSTest. Actually, we probably should use whatever monogame uses.

A separate repo for demos is a good idea, but we'll also need to keep a kind of sandbox in the main repo for building features I think.

`Contains` works better I think simply because it matches the existing MonoGame API. 

Anyway, I'm going to bed. It's really late here.

**Comment by WardBenjamin** [comment]
Night!

I'm pretty sure that MonoGame uses NUnit, though I can check that out in a second. I'll edit this post with more information.

**Comment by craftworkgames** [comment]
> Yep, MonoGame definitely uses NUnit, so I would suggest going with that. That decision is up to you though.

I think a contributor to this project might also be a contributor to MonoGame. I could be wrong about that but it's a good a reason as any. I don't know of any reason not to use NUnit.

**Comment by craftworkgames** [comment]
I've added NUnit to the project with 1 test. Unfortunately there's not a lot we can actually test at this stage because it's impossible to mock / substitute MonoGame's `GraphicsDevice`. We *could* make the code more testable, but I'm not sure it's worth it because it will also make it more complicated.

Anyway, here's the first test run on the build system.
http://build.craftworkgames.com/viewLog.html?buildId=229&tab=buildResultsDiv&buildTypeId=MonoGameExtended_CI

How's the camera stuff going?

**Comment by WardBenjamin** [comment]
I currently am doing some testing of the bounding frustrum; the other methods should be trivial to implement. I'll probably submit a pull request either today or tomorrow based on when I have internet access, since I'm traveling today.

**Comment by WardBenjamin** [comment]
So, I've been doing some testing and I just ran the Game1 executable for the first time. Even without my changes, it seems to be overflowing the stack (literally, a StackOverflowException). It's occurring in BoxingViewportAdapter supposedly, but I'll hunt around for it a bit.

**Comment by craftworkgames** [comment]
@WardBenjamin No problem. I remember the stack overflow exception from a few days ago. My fault for committing it. It's just a coincidence that you happened to fork it at that time.

Looking forward to the pull request. Once it's done I'm going to restructure the project a bit (renames some folders, move some classes into namespaces)

**Comment by craftworkgames** [comment]
Oh btw.. I've raised another issue over here #2 if you'd like to comment.

**Comment by WardBenjamin** [comment]
@craftworkgames I've added a comment on #2, as well as reaching out to a guy that I know on GameDev.StackExchange for any further ideas that he may have (he used Monogame for a long time before switching to Unity, because Monogame lacked features such as the ones you are suggesting).

**Comment by WardBenjamin** [comment]
@craftworkgames How do you feel about region directives? I personally like to keep all overloads of a method inside a region if there are more than a few just so I can collapse them, but I would like some verification that that is alright with you.

**Comment by craftworkgames** [comment]
@WardBenjamin I'm really not a big fan of regions to be honest. I think they clutter up the code and make it harder to read.

**Comment by WardBenjamin** [comment]
Alright, that's fine. I'll remove them. I've just about finished with the basic `Contains` methods, the rectangle one is proving hard to find an elegant solution.

**Comment by WardBenjamin** [comment]
I think this can be closed as of pull request #3 and #4, once it is merged.

**Comment by WardBenjamin** [comment]
As a note, something that could be considered for the Contains(Rectangle) overload, Dustin Horne did some experimentation with that topic, which can be read here: http://www.dustinhorne.com/post/2011/10/18/XNA-Terrain-with-LOD-Part-7-Better-Culling-With-2D-Shapes

**Comment by craftworkgames** [comment]
@WardBenjamin Thanks for working on this.

If it's okay with you I'm going to do some project restructuring later tonight. I'll be moving files around and renaming folders so it could turn into a nasty merge if you've got any uncommitted changes? Just wondering if you can hold off making any more changes for the half day or so?

**Comment by craftworkgames** [comment]
I added a few unit tests around the new camera functionality and found some bugs. There was also a bug in another class that I picked up along the way :dizzy_face: Anyway.. at least the unit tests are proving to have value.

**Comment by WardBenjamin** [comment]
@craftworkgames yeah, I can definitely hold off on changes for a while. I leave for vacation again on Tuesday so I'm spending quite a bit of time packing :angry:

What were some of the camera bugs?

**Comment by craftworkgames** [comment]
@WardBenjamin Don't worry about it. We all make mistakes like that. You were probably just excited to get your changed merged :smile: 

I switched the project from OpenGL to DirectX (because @roastedamoeba on twitter said that it would make the unit testing easier). He was right, but I also found the DirectX version has some bugs when setting the `GraphicsDevice.Viewport` property. I'm looking at that now.

**Comment by WardBenjamin** [comment]
:grinning: 

That's interesting that the DX code has some bugs like that. I know that the half-pixel offest thing for the projection matrix actually is native for DX; It's in the GL SpriteBatch to match DX's native.... Did that effect the test?

**Comment by craftworkgames** [comment]
Hard to say. I never actually got the test working with GL.

**Comment by craftworkgames** [comment]
> yeah, I can definitely hold off on changes for a while. I leave for vacation again on Tuesday so I'm spending quite a bit of time packing :angry:

@WardBenjamin I've finished the folder restructure. I recommend deleting your local repo and getting it fresh because it seems to leave old folders around. I probably didn't do it quite right.

In particular, I renamed the root `MonoGame.Extended` folder to `Source` and when I got latest version on another computer it left the old `MonoGame.Extended` folder and I had to delete it manually.
